### PR TITLE
Week based parsing for ingest date processor

### LINKED ISSUE #58479: Date processor incorrectly parses week based dates with java.time
ingest `date time` processor is incorrectly parsing week based date formats.
raised by a user on a discuss forum https://discuss.elastic.co/t/elastic-joda-to-java-time-update-illegal-pattern-component-uuuu/237830
It affects v8, v7 and 6.8 versions
class: https://github.com/elastic/elasticsearch/blob/master/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateFormat.java#L82

reproduce step in 6.8
```
curl --request PUT \
  --url http://localhost:9200/_ingest/pipeline/date_pipeline \
  --header 'authorization: Basic ZWxhc3RpYzpwYXNzd29yZA==' \
  --header 'content-type: application/json' \
  --data '{
  "description": "Pipeline for parsing date",
  "processors": [
    {
      "date": {
        "field": "createdTime",
        "formats": [
         "8YYYY-ww"
        ]
      }
    }
  ]
}'
```

simulate
```
curl --request POST \
  --url http://localhost:9200/_ingest/pipeline/_simulate \
  --header 'authorization: Basic ZWxhc3RpYzpwYXNzd29yZA==' \
  --header 'content-type: application/json' \
  --data '{
  "pipeline": {
    "processors": [
      {
        "pipeline": {
          "name": "date_pipeline"
        }
      }
    ]
  },
  "docs": [
    {
      "_source": {
 				 "createdTime": "2020-33"
      }
    }
  ]
}'
```

incorrect response
```
{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_type": "_type",
        "_id": "_id",
        "_source": {
          "createdTime": "2020-33",
          "@timestamp": "2020-01-01T00:00:00.000Z"
        },
        "_ingest": {
          "timestamp": "2020-06-24T08:12:19.899156Z"
        }
      }
    }
  ]
}
```

when ingesting it fails with
```
curl --request PUT \
  --url 'http://localhost:9200/test1/_doc/1?pipeline=date_pipeline' \
  --header 'authorization: Basic ZWxhc3RpYzpwYXNzd29yZA==' \
  --header 'content-type: application/json' \
  --data '{
  "createdTime": "2020-33"
}'
```
```
{
  "error": {
    "root_cause": [
      {
        "type": "mapper_parsing_exception",
        "reason": "failed to parse field [createdTime] of type [date] in document with id '1'. Preview of field's value: '2020-33'"
      }
    ],
    "type": "mapper_parsing_exception",
    "reason": "failed to parse field [createdTime] of type [date] in document with id '1'. Preview of field's value: '2020-33'",
    "caused_by": {
      "type": "date_time_exception",
      "reason": "Invalid value for MonthOfYear (valid values 1 - 12): 33"
    }
  },
  "status": 400
}
```

it works for joda time implementation in 6.8
```
curl --request PUT \
  --url http://localhost:9200/_ingest/pipeline/date_pipeline \
  --header 'authorization: Basic ZWxhc3RpYzpwYXNzd29yZA==' \
  --header 'content-type: application/json' \
  --data '{
  "description": "Pipeline for parsing date",
  "processors": [
    {
      "date": {
        "field": "createdTime",
        "formats": [
         "yyyy-ww"
        ]
      }
    }
  ]
}'
```

```
curl --request POST \
  --url http://localhost:9200/_ingest/pipeline/_simulate \
  --header 'authorization: Basic ZWxhc3RpYzpwYXNzd29yZA==' \
  --header 'content-type: application/json' \
  --data '{
  "pipeline": {
    "processors": [
      {
        "pipeline": {
          "name": "date_pipeline"
        }
      }
    ]
  },
  "docs": [
    {
      "_source": {
 				 "createdTime": "2020-33"
      }
    }
  ]
}'
```
correct response - see 
```
{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_type": "_type",
        "_id": "_id",
        "_source": {
          "createdTime": "2020-33",
          "@timestamp": "2020-08-10T00:00:00.000Z"
        },
        "_ingest": {
          "timestamp": "2020-06-24T08:44:48.250843Z"
        }
      }
    }
  ]
}
``

### PR DESCRIPTION
**Pull request by pgomulka**
Date processor was incorrectly parsing week based dates as when a
weekbased year was provided ingest module was thinking year was not
on a date and was trying to applying the logic for `dd/MM` type of
dates.
For testing ingest has to have access to ISOCalendarDataProvider in
order to correctly calculate week based dates with ISO rules.
Date Processor is also allowing users to specify locale parameter. It
should be taken into account when parsing dates - currently only used
for formatting. If someone specifies 'en-us' locale, then calendar data
rules for that locale should be used.
The exception is iso8601 format. If someone is using that format, 
then locale should not override calendar data rules.
closes https://github.com/elastic/elasticsearch/issues/58479

### CODE REVIEW

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateFormat.java [hunk 1]
```diff
@@ -45,8 +46,13 @@
     Iso8601 {
         @Override
         Function<String, ZonedDateTime> getFunction(String format, ZoneId timezone, Locale locale) {
-            return (date) -> DateFormatters.from(DateFormatter.forPattern("iso8601").parse(date), timezone)
-                                            .withZoneSameInstant(timezone);
+            //
```
Comment (rjernst) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateFormat.java:49]: unfinished comment?

File: modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/DateFormatTests.java [hunk 2]
```diff
@@ -126,4 +143,6 @@ public void testFromString() {
         assertThat(DateFormat.fromString("tai64n"), equalTo(DateFormat.Java));
         assertThat(DateFormat.fromString("prefix-" + randomAlphaOfLengthBetween(1, 10)), equalTo(DateFormat.Java));
     }
+
```
Comment (rjernst) [inline modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/DateFormatTests.java:146]: nit: extra newlines?

File: modules/ingest-common/build.gradle [hunk 3]
```diff
@@ -25,6 +25,7 @@ esplugin {
 
 dependencies {
   compileOnly project(':modules:lang-painless')
+  compile project(':libs:elasticsearch-core')
```
Comment (rjernst) [inline modules/ingest-common/build.gradle:28]: Why is this necessary? modules depend on server, implicitly, so we should be able to use the date classes moved there from server, not need them in core lib

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateFormat.java (hunk 1 above)
Comment (pgomulka) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DateFormat.java:49]: will remove

File: modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/DateFormatTests.java (hunk 2 above)
Comment (pgomulka) [reply modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/DateFormatTests.java:146]: will remove

File: modules/ingest-common/build.gradle (hunk 3 above)
Comment (pgomulka) [reply modules/ingest-common/build.gradle:28]: I had problem accessing `IsoCalendarDataProvider` from server when running ingest yml tests.
Maybe adding this as test only? But then again it has to be available when running a cluster for yml test.
Any ideas?

**Comment by pgomulka** [comment]
ok to test

File: modules/ingest-common/build.gradle (hunk 3 above)
Comment (pgomulka) [reply modules/ingest-common/build.gradle]: interesting. I cannot reproduce that failure now.  I can see from dependency tree that server is indeed on a test compile classpath. Not sure how was this failing for me before..
Reverted back the change moving `IsoCalendarDataProvider`

**Review by rjernst — APPROVED** [review]
LGTM

**Pull request merged**